src/storage/src/vector/mod.rs

```diff
@@ -0,0 +1,537 @@
+        // TODO: use some library with simd support
+        let len = self.0.0.len();
+        assert_eq!(len, other.0.len());
+        // In this implementation, we don't take the square root to avoid unnecessary computation, because
+        // we only want comparison rather than the actual distance.
+        (0..len)
+            .map(|i| {
+                let diff = self.0.0[i] - other.0[i];
+                diff.abs()
+            })
+            .sum()
+    }
+}
+
+pub trait MeasureDistance<'a> {
+    fn measure(&self, other: VectorRef<'_>) -> VectorDistance;
+}
+
+pub struct L2Distance;
+
+pub struct L2DistanceMeasure<'a>(VectorRef<'a>);
+
+impl MeasureDistanceBuilder for L2Distance {
+    type Measure<'a> = L2DistanceMeasure<'a>;
+
+    fn new(target: VectorRef<'_>) -> Self::Measure<'_> {
+        L2DistanceMeasure(target)
+    }
+}
+
+impl<'a> MeasureDistance<'a> for L2DistanceMeasure<'a> {
+    fn measure(&self, other: VectorRef<'_>) -> VectorDistance {
+        // TODO: use some library with simd support
+        let len = self.0.0.len();
+        assert_eq!(len, other.0.len());
+        // In this implementation, we don't take the square root to avoid unnecessary computation, because
+        // we only want comparison rather than the actual distance.
+        (0..len).map(|i| (self.0.0[i] - other.0[i]).powi(2)).sum()
+    }
+}
+
+pub struct KlDivergenceDistance;
+pub struct KlDivergenceDistanceMeasure<'a>(VectorRef<'a>);
+
+impl MeasureDistanceBuilder for KlDivergenceDistance {
+    type Measure<'a> = KlDivergenceDistanceMeasure<'a>;
+
+    fn new(target: VectorRef<'_>) -> KlDivergenceDistanceMeasure<'_> {
+        KlDivergenceDistanceMeasure(target)
+    }
+}
+
+impl<'a> MeasureDistance<'a> for KlDivergenceDistanceMeasure<'a> {
+    fn measure(&self, other: VectorRef<'_>) -> VectorDistance {
+        // TODO: use some library with simd support
+        let len = self.0.0.len();
+        assert_eq!(len, other.0.len());
+        (0..len)
+            .map(|i| {
+                let target_prob = self.0.0[i];
+                target_prob * (target_prob / other.0[i]).ln()
+            })
+            .sum()
+    }
+}
+
+pub struct CosineDistance;
+pub struct CosineDistanceMeasure<'a> {
+    target: VectorRef<'a>,
+    magnitude: VectorItem,
+}
+
+impl MeasureDistanceBuilder for CosineDistance {
+    type Measure<'a> = CosineDistanceMeasure<'a>;
+
+    fn new(target: VectorRef<'_>) -> Self::Measure<'_> {
+        let magnitude = target.magnitude();
+        CosineDistanceMeasure { target, magnitude }
+    }
+}
+
+impl<'a> MeasureDistance<'a> for CosineDistanceMeasure<'a> {
+    fn measure(&self, other: VectorRef<'_>) -> VectorDistance {
+        // TODO: use some library with simd support
+        let len = self.target.0.len();
+        assert_eq!(len, other.0.len());
+        let magnitude_mul = other.magnitude() * self.magnitude;
+        (0..len)
+            .map(|i| self.target.0[i] * other.0[i] / magnitude_mul)
```

Comment:
Cosine Distance appears to be 1 - xxx. Could this reverse the sorting results here?

Reply:
Seems not. Cosine distance is the cosine value on the angle of two vectors. You may refer to https://en.wikipedia.org/wiki/Cosine_similarity

When the two vectors are normalized, the cosine simarility is equal to inner product of them.

Reply:
Oh wait. Indeed, when two vector is the same, their cosine similarity is 1, but they should be at the nearest distance. It's true that we should use the `1-similarity` as the distance. Thanks for pointing it out.